### PR TITLE
NIFI-12290 Migrate from Quartz to Spring for Cron Expressions

### PR DESCRIPTION
**Pull request by exceptionfactory**
# Summary

[NIFI-12290](https://issues.apache.org/jira/browse/NIFI-12290) Migrates from Quartz to Spring Framework [CronExpression](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/scheduling/support/CronExpression.html) for Cron scheduling of components.

The Quartz library relies on `java.util.Date` and has several legacy dependencies including HikariCP and C3P0 which are not necessary for project operations.

The Spring Framework 5.3 [introduced new Cron Expression](https://spring.io/blog/2020/11/10/new-in-spring-5-3-improved-cron-expressions) capabilities based on `java.time` classes that follow standard Unix crontab conventions.

The differences from Quartz are minimal, but include the removal of the optional Year field, and the change from 1-7 to 0-7 for numeric Day of Week references.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by mosermw** [comment]
We should remove Quartz and c3p0 from the various NOTICE files that reference it.

File: nifi-docs/src/main/asciidoc/user-guide.adoc [hunk 1]
```diff
@@ -577,10 +577,8 @@ The first configuration option is the Scheduling Strategy. There are three possi
 |1-12 or JAN-DEC
 
 |Day of Week
-|1-7 or SUN-SAT
+|0-6 or SUN-SAT (0 or 7 is SUN)
```
Comment (mosermw) [inline nifi-docs/src/main/asciidoc/user-guide.adoc:580]: It's a bit odd that the range is 0-6 but 7 = SUN? Is there a better way of expressing this?

**Comment by mosermw** [comment]
I tried this with GenerateFlowFile and CRON driven scheduling with default "* * * * * ?" and I got this validation error:
`14:00:20 UTC  ERROR GenerateFlowFile[id=00000000-0000-0000-0000-000000000000] Failed to properlt initialize Processor. If still scheduled to run, NiFi will attempt to initialize and run the Processor again after the 'Administrative Yield Duration' has elapsed. Failure is due to java.tmie.temporal.UnsupportedTemporalTypeException: Unsupported field: DayOfWeek: java.time.temporal.UnsupportedTemporalTypeException: Unsupported field: DayOfWeek`

File: nifi-docs/src/main/asciidoc/user-guide.adoc (hunk 1 above)
Comment (exceptionfactory) [reply nifi-docs/src/main/asciidoc/user-guide.adoc]: Thanks for highlighting that example, I changed the documentation to read `0-7` with `(0 or 7 is Sunday)`. That follows the Cron Expression documentation from the Spring Framework, and other places. Open to other suggestions if you think additional clarification is needed.

**Comment by mosermw** [comment]
+1 will merge to main.  It's always good to see improvements that reduce NiFi dependencies.

@exceptionfactory Does the CRON format change for DayOfWeek (SUN-SAT changes from 1-7 to 0-6) warrant a 2.0 Migration Guide note?

**Comment by exceptionfactory** [comment]
Thanks for reviewing and merging @mosermw!

I added a new section for [CRON Scheduling Changes](https://cwiki.apache.org/confluence/display/NIFI/Migrating+Deprecated+Components+and+Features+for+2.0.0#MigratingDeprecatedComponentsandFeaturesfor2.0.0-CRONSchedulingChanges) to the 2.0 migration documentation, which will be linked from the regular Migration Guidance page when releasing.

**Pull request closed without merging**